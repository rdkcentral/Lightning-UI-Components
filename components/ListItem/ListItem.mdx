import { Preview, Story } from '@storybook/addon-docs/blocks';
// uncomment the line below and edit the path to point to your component
// import * as stories from './ListItem.stories';

# ListItem

Component to be displayed in a list

## Usage

`ListItem` can display a `title`, `subtitle`, and `icon`:

```js
import { ListItem } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      ListItem: {
        type: ListItem,
        title: 'List Item',
        subtitle: 'List item metadata',
        icon: 'path/to/icon.png'
      }
    };
  }
};
```

<Preview>
  <Story id="listitem--basic" />
</Preview>

The expected icon size is 40x40.

### Background

For a transparent background, set `background: 'float'`. The default value is `fill`.

### Sizing

There are three sizing options, `small`, `large`, and any positive integer. The default value is `large`. These options only affect width, the height is fixed at 88px.

### Focus

When focused, the background/foreground colors are inverted.

### Icons

Adding icons to a `ListItem` is optional, and they can be added one of two ways:

**Pass a url path**

```js
{
  type: ListItem,
  icon: 'path/to/icon.png'
}
```

**Pass a list of paths**

```js
{
  type: ListItem,
  icon: ['path/to/icon.png', 'path/to/other/icon.png']
}
```

If multiple icons are passed, they will be laid out horizontally in order.

<Preview>
  <Story id="listitem--multiple-icons" />
</Preview>

## API

### Properties

name|type|readonly|description
--|--|--|--
background|'fill'\|'float'|false|color of `ListItem` container.
icon|string\|string[]|false|path to icon image (or list of paths)
size|'small'\|'large'\|number|false|width of the `ListItem` container
subtitle|string|false|item metadata info goes here
title|string|false|main item info goes here

### Methods

#### _handleEnter

Can be overridden to apply handling for the <kbd>Enter</kbd> key
