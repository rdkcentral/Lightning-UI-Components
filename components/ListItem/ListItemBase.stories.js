import lng from 'wpe-lightning';
import { boolean, radios, withKnobs } from '@storybook/addon-knobs';

import { ListItemBase } from '.';
import mdx from './ListItem.mdx';
import { makeOptions } from '../../.storybook/utils';

export default {
  title: 'ListItemBase',
  component: ListItemBase,
  decorators: [withKnobs],
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const Basic = () =>
  class Basic extends lng.Component {
    static _template() {
      return {
        ListItem: {
          type: ListItemBase,
          size: radios('Size', makeOptions('Large', 'Small'), 'large'),
          background: radios('Background', makeOptions('Fill', 'Float'), 'fill')
        }
      };
    }

    _init() {
      this.tag('ListItem')._Left.patch({
        w: 50,
        h: 50,
        rect: true
      });
      this.tag('ListItem')._Right.patch({
        w: 50,
        h: 50,
        rect: true
      });
    }

    _getFocused() {
      if (boolean('Focused', false)) {
        return this.tag('ListItem');
      }
    }
  };

export const FullWidth = () =>
  class Basic extends lng.Component {
    static _template() {
      return {
        ListItem: {
          type: ListItemBase,
          leftSlot: {
            h: 50,
            rect: true,
            flexItem: { grow: 1 }
          }
        }
      };
    }

    _init() {
      this.tag('ListItem')._Left.patch({
        h: 50,
        rect: true,
        flexItem: { grow: 1 }
      });
    }
  };

export const Stacked = () =>
  class Stacked extends lng.Component {
    static _template() {
      return {
        ListItem: { type: ListItemBase }
      };
    }

    _init() {
      const slot = {
        Top: {
          h: 30,
          w: 80,
          rect: true,
          flexItem: { margin: 1 }
        },
        Bottom: {
          h: 30,
          w: 80,
          rect: true,
          flexItem: { margin: 1 }
        }
      };

      this.tag('ListItem')._Left.patch({ ...slot });
      this.tag('ListItem')._Right.patch({ ...slot });
    }
  };

export const Text = () =>
  class Text extends lng.Component {
    static _template() {
      return {
        ListItem: { type: ListItemBase }
      };
    }

    _init() {
      this.tag('ListItem')._Left.patch({
        Text: {
          text: {
            fontSize: 28,
            h: 40,
            text: 'Hello world'
          }
        }
      });
    }
  };

export const StackedText = () =>
  class StackedText extends lng.Component {
    static _template() {
      return {
        ListItem: { type: ListItemBase }
      };
    }

    _init() {
      this.tag('ListItem')._Left.patch({
        Text1: {
          text: {
            fontSize: 28,
            text: `This is a title`
          }
        },
        Text2: {
          alpha: 0.7,
          text: {
            fontSize: 22,
            text: `this is a subtitle`
          }
        }
      });
    }
  };
