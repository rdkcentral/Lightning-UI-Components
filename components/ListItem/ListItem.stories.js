import lng from 'wpe-lightning';

import ListItem from '.';
import mdx from './ListItem.mdx';
import icon from '../../assets/images/ic_lightning_white_32.png';

export default {
  title: 'ListItem',
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const Basic = args =>
  class Basic extends lng.Component {
    static _template() {
      return {
        ListItem: {
          type: ListItem,
          title: args.title,
          subtitle: args.subtitle,
          collapse: args.collapse,
          icon,
          size: args.size,
          backgroundType: args.backgroundType
        }
      };
    }

    _getFocused() {
      if (args.focused) {
        return this.tag('ListItem');
      }
    }
  };
Basic.args = {
  backgroundType: 'fill',
  focused: false,
  title: 'List Item',
  subtitle: 'List item metadata',
  collapse: false,
  size: 'small'
};
Basic.argTypes = {
  size: {
    control: {
      type: 'radio',
      options: ['small', 'large']
    }
  },
  backgroundType: {
    control: {
      type: 'radio',
      options: ['fill', 'float', 'ghost']
    }
  }
};
Basic.parameters = {
  argActions: {
    focused: (isFocused, component) => {
      component._getFocused = isFocused
        ? () => component.tag('ListItem')
        : () => {};
      component._refocus();
    }
  }
};

export const KeyHandling = () =>
  class KeyHandling extends lng.Component {
    static _template() {
      return {
        flex: {
          direction: 'column'
        },
        ListItem: {
          type: ListItem,
          size: 'small',
          title: 'Press Enter'
        },
        HiddenText: {
          alpha: 0,
          flexItem: { marginTop: 100 },
          text: {
            text: 'Great job!'
          }
        }
      };
    }

    _init() {
      const HiddenText = this.tag('HiddenText');
      this.tag('ListItem').onEnter = () => {
        HiddenText.smooth = { alpha: Number(!HiddenText.alpha) };
      };
    }

    _getFocused() {
      return this.tag('ListItem');
    }
  };

export const MultipleIcons = () =>
  class MultipleIcons extends lng.Component {
    static _template() {
      return {
        ListItem: {
          type: ListItem,
          title: 'List Item',
          subtitle: 'List item metadata',
          icon: [icon, icon],
          size: 'small',
          backgroundType: 'fill'
        }
      };
    }
  };
