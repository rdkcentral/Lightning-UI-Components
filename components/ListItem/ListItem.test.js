import ListItem, { ListItemBase } from '.';
import { getHexColor } from '../Styles';
import TestUtils from '../lightning-test-utils';

const createListItemBase = TestUtils.makeCreateComponent(ListItemBase);
const createListItem = TestUtils.makeCreateComponent(ListItem);
const icon = TestUtils.pathToDataURI('assets/images/ic_lightning_white_32.png');

describe('ListItemBase', () => {
  let listItemBase, testRenderer;
  beforeEach(() => {
    [listItemBase, testRenderer] = createListItemBase();
    testRenderer.update();
  });
  afterEach(() => {
    listItemBase = null;
    testRenderer = null;
  });

  it('should render', () => {
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the correct height', () => {
    expect(listItemBase._Container.finalH).toBe(88);
  });

  it('renders small size', () => {
    [listItemBase, testRenderer] = createListItemBase({ size: 'small' });
    testRenderer.update();
    expect(listItemBase._Container.finalW).toBe(410);
  });

  it('renders large size', () => {
    [listItemBase, testRenderer] = createListItemBase({ size: 'large' });
    testRenderer.update();
    expect(listItemBase._Container.finalW).toBe(860);
  });

  it('renders custom size', () => {
    [listItemBase, testRenderer] = createListItemBase({ w: 500 });
    testRenderer.update();
    expect(listItemBase._Container.finalW).toBe(500);
  });

  describe('focus', () => {
    it('has a default unfocus background', () => {
      expect(listItemBase._Container.color).toBe(getHexColor('141417'));
    });
    it('supports background props', () => {
      [listItemBase, testRenderer] = createListItemBase({ background: 'fill' });
      testRenderer.update();
      expect(listItemBase._Container.color).toBe(getHexColor('141417'));

      [listItemBase, testRenderer] = createListItemBase({
        background: 'float'
      });
      testRenderer.update();
      expect(listItemBase._Container.color).toBe(0);
    });
    it('has a focus background', () => {
      listItemBase._focus();
      testRenderer.update();
      expect(listItemBase._Container.color).toBe(getHexColor('ECECF2'));
    });
    it('resets colors on unfocus', () => {
      listItemBase._focus();
      testRenderer.update();

      listItemBase._unfocus();
      testRenderer.update();
      expect(listItemBase._Container.color).toBe(getHexColor('141417'));
    });
  });
});

describe('ListItem', () => {
  let listItem, testRenderer;

  beforeEach(() => {
    [listItem, testRenderer] = createListItem();
    testRenderer.update();
  });

  afterEach(() => {
    listItem = null;
    testRenderer = null;
  });

  it('should render', () => {
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a title', () => {
    [listItem, testRenderer] = createListItem({
      title: 'My Title'
    });
    testRenderer.update();
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
    expect(listItem._Title.text.text).toEqual(listItem.title);
  });

  it('should render a subtitle', () => {
    [listItem, testRenderer] = createListItem({
      subtitle: 'My Subtitle'
    });
    testRenderer.update();
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
    expect(listItem._Subtitle.text.text).toEqual(listItem.subtitle);
  });

  it('should render a title and subtitle together', () => {
    [listItem, testRenderer] = createListItem({
      title: 'My Title',
      subtitle: 'My Subtitle'
    });
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render an icon', () => {
    [listItem, testRenderer] = createListItem({
      icon: icon
    });
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
    expect(listItem._Icon.src).toEqual(listItem.icon);
  });

  describe('focus', () => {
    beforeEach(() => {
      [listItem, testRenderer] = createListItem({
        title: 'My Title',
        subtitle: 'My Subtitle',
        icon: icon
      });
      listItem._focus();
      testRenderer.update();
    });
    it('focused items transition color', () => {
      expect(listItem._Title.color).toEqual(0xff000000);
      expect(listItem._Subtitle.color).toEqual(0xff000000);
      expect(listItem._Icon.color).toEqual(0xff000000);
    });
    it('unfocused items transition color', () => {
      listItem._unfocus();
      testRenderer.update();

      expect(listItem._Title.color).toEqual(0xffffffff);
      expect(listItem._Subtitle.color).toEqual(0xffffffff);
      expect(listItem._Icon.color).toEqual(0xffffffff);
    });
  });
});
