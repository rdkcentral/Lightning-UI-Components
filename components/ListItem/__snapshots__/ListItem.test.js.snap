// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem should render 1`] = `
Object {
  "Component": Object {
    "Container": Object {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Left": Object {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "id": 74,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Left",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Left",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Right": Object {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "id": 75,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Right",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Right",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": "ff141417",
      "enabled": true,
      "flex": true,
      "flexItem": true,
      "h": 0,
      "id": 73,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Container",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Container",
      ],
      "texture": Object {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;

exports[`ListItem should render a subtitle 1`] = `
Object {
  "Component": Object {
    "Container": Object {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Left": Object {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Subtitle": Object {
              "active": true,
              "alpha": 0.8,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": true,
              "h": 0,
              "id": 97,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Subtitle",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": Array [
                "Subtitle",
              ],
              "texture": Object {
                "fontFace": "XfinityStandardMedium",
                "fontSize": 24,
                "h": 32,
                "text": "My Subtitle",
                "type": "TextTexture",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "id": 95,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Left",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Left",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Right": Object {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "id": 96,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Right",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Right",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": "ff141417",
      "enabled": true,
      "flex": true,
      "flexItem": true,
      "h": 0,
      "id": 94,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Container",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Container",
      ],
      "texture": Object {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;

exports[`ListItem should render a title 1`] = `
Object {
  "Component": Object {
    "Container": Object {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Left": Object {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Title": Object {
              "active": true,
              "alpha": 0.95,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": true,
              "h": 0,
              "id": 86,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Title",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": Array [
                "Title",
              ],
              "texture": Object {
                "fontFace": "XfinityStandardBold",
                "fontSize": 28,
                "h": 40,
                "text": "My Title",
                "type": "TextTexture",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "id": 84,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Left",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Left",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Right": Object {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "id": 85,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Right",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Right",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": "ff141417",
      "enabled": true,
      "flex": true,
      "flexItem": true,
      "h": 0,
      "id": 83,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Container",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Container",
      ],
      "texture": Object {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;

exports[`ListItem should render a title and subtitle together 1`] = `
Object {
  "Component": Object {
    "Container": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Left": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Subtitle": Object {
              "active": false,
              "alpha": 0.8,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": true,
              "h": 0,
              "id": 109,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Subtitle",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": Array [
                "Subtitle",
              ],
              "texture": Object {
                "fontFace": "XfinityStandardMedium",
                "fontSize": 24,
                "h": 32,
                "text": "My Subtitle",
                "type": "TextTexture",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
            "Title": Object {
              "active": false,
              "alpha": 0.95,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": true,
              "h": 0,
              "id": 108,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Title",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": Array [
                "Title",
              ],
              "texture": Object {
                "fontFace": "XfinityStandardBold",
                "fontSize": 28,
                "h": 40,
                "text": "My Title",
                "type": "TextTexture",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "id": 106,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Left",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Left",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Right": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "id": 107,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Right",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Right",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": "ff141417",
      "enabled": true,
      "flex": true,
      "flexItem": true,
      "h": 0,
      "id": 105,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Container",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Container",
      ],
      "texture": Object {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;

exports[`ListItem should render an icon 1`] = `
Object {
  "Component": Object {
    "Container": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Left": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "id": 118,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Left",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Left",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Right": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Icon": Object {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": true,
              "h": 40,
              "id": 120,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Icon",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": Array [
                "Icon",
              ],
              "texture": Object {
                "src": "data:image/jpg;base64, iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAABTUlEQVRYR8WX0VEDMQxEdyuADggVhFQAVEALdJBQASmBVEALdABUAB0AHZAKllGGMDc3Z/tGRrZ+7uNi+2mtlS5E52DN+ZJOSX7X7OEGkHQL4JGkew8Ddy22zAF8ADAFXHscVXMtlrQFcH/IoDWApMVv9ockegA8Abj5k7ClApKuADwPq76pApLeASy7ABxtN/Z8EwWGtusF8ABgPdXxwhUY2665ApJeAFxm+r29T8UngB1JK97JyHbCKds5Bo8Nq/PU0CoBWAZnjkPHS1YpFZIAki4AvP3D4V8krX27rsDkO6mEuCaZrJPSFZgKNvftmYpcgb6StPadDNc4Hu4mSZn9rfisjroAmP02peuLUmAPYDHnezEK4I6kte9iRABkbTcmigDI2i4aoGi7aICi7SIBZtkuCmC27SIAbF5s59ouAqDqD2q1DYudpvCDH6f0jiFcMsG+AAAAAElFTkSuQmCC",
                "type": "ImageTexture",
              },
              "visible": true,
              "w": 40,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "id": 119,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Right",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Right",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": "ff141417",
      "enabled": true,
      "flex": true,
      "flexItem": true,
      "h": 0,
      "id": 117,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Container",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Container",
      ],
      "texture": Object {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;

exports[`ListItemBase should render 1`] = `
Object {
  "Component": Object {
    "Container": Object {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Left": Object {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "id": 4,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Left",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Left",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Right": Object {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "id": 5,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Right",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Right",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": "ff141417",
      "enabled": true,
      "flex": true,
      "flexItem": true,
      "h": 0,
      "id": 3,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Container",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Container",
      ],
      "texture": Object {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
